SWUTILS-931: Add backup alternatives for depreciated Linux utilities

dchadwic-xilinx opened the pull request:
This patch adds a backup in the case `netstat`, `ifconfig`, `arp` and/or `route` are not available.  These utils are being depreciated and will be not be automatically installed in future Linux OS releases.

I have tested this on my systems but I would like to get hold of the newest RHEL 9.X or candidate 10 to test.

amitgu-xilinx commented:
"ip route show" and  "ip -s -d addr show" commands are already used in sfreport.  So might not be required to be reported twice.

amitgu-xilinx commented:
Also worth updating the headline from "ifconfig" to more generic e.g" Network Configuration"

amitgu-xilinx commented:
Might be worth reporting both outputs. So some people prefer old and other new commands

dchadwic-xilinx commented:
@amitgu-xilinx @ibeecraft-amd 
>My preference is to change the order to pull in the newer commands by default and revert to the older commands if required (on systems with ip for example). Others may prefer the formatting of the older ones but if these commands are scheduled to be removed from future RHEL we should get used to their output.

Agree with this. The new commands now appear first.

> "ip route show" and "ip -s -d addr show" commands are already used in sfreport. So might not be required to be reported twice.

Addressed this by removing the repeats of `ip -4/-6 route show`.

> Might be worth reporting both outputs. So some people prefer old and other new commands

I have addressed this by keeping both outputs. However, in the case the new output is available the old output is hidden by default. The reason for this is to reduce the increased scrolling time.

amitgu-xilinx commented:
[AMD Official Use Only - AMD Internal Distribution Only]

@Chadwick, ***@***.***>

I have addressed this by keeping both outputs. However, in the case the new output is available the old output is hidden by default. The reason for this is to reduce the increased scrolling time.
Jay’s sfreport reports still rely on the old data, so until he has a fix for this, we should keep the output visible. Or discuss with Jay.


From: Daniel Chadwick ***@***.***>
Sent: Thursday, September 26, 2024 3:59 PM
To: Xilinx-CNS/sfreport ***@***.***>
Cc: Gupta, Amit UK ***@***.***>; Mention ***@***.***>
Subject: Re: [Xilinx-CNS/sfreport] SWUTILS-931: Add backup alternatives for depreciated Linux utilities (PR #25)

Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


@amitgu-xilinx<https://github.com/amitgu-xilinx> @ibeecraft-amd<https://github.com/ibeecraft-amd>

My preference is to change the order to pull in the newer commands by default and revert to the older commands if required (on systems with ip for example). Others may prefer the formatting of the older ones but if these commands are scheduled to be removed from future RHEL we should get used to their output.

Agree with this. The new commands now appear first.

"ip route show" and "ip -s -d addr show" commands are already used in sfreport. So might not be required to be reported twice.

Addressed this by removing the repeats of ip -4/-6 route show.

Might be worth reporting both outputs. So some people prefer old and other new commands

I have addressed this by keeping both outputs. However, in the case the new output is available the old output is hidden by default. The reason for this is to reduce the increased scrolling time.

—
Reply to this email directly, view it on GitHub<https://github.com/Xilinx-CNS/sfreport/pull/25#issuecomment-2377218481>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOMZGGJUOWAA2C6AQLSF2QTZYQOJVAVCNFSM6AAAAABOX646LKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNZXGIYTQNBYGE>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

dchadwic-xilinx commented:
> Jay’s sfreport reports still rely on the old data, so until he has a fix for this, we should keep the output visible. Or discuss with Jay.
I assume the miner uses the raw html in which case I would guess the lines being visible in the HTML viewer doesn't matter.
@jjay-xilinx any comment?

jjay-xilinx commented:
> I assume the miner uses the raw html in which case I would guess the lines being visible in the HTML viewer doesn't matter.
Correct.

It'll be a while before my miner encounters a customer sfreport in this new format, so plenty of time for me to fix and work around it.

dchadwic-xilinx commented:
I have reviewed and tested, and these changes do not impact the utility of the dataminer. This is something I will keep in mind when reviewing future sfreport changes. We must be mindful of changes which may cause the dataminer to false positive/negative. 

>>Might be worth reporting both outputs. So some people prefer old and other new commands
>
>I have addressed this by keeping both outputs. However, in the case the new output is available the old output is hidden by default. The reason for this is to reduce the increased scrolling time.

In hindsight, this is the only correct decision. It may be in future that we have to do some magic with the miner to unify the overlap between these fields for comparison.

dchadwic-xilinx commented:
This is ready for review again. I updated the `ip route show table all` output to be tabluated:
![image](https://github.com/user-attachments/assets/0068d510-964b-43d1-ba60-2c10537a5e41)
This is easier to parse with the miner if required in furture.

dchadwic-xilinx commented:
> May be a good time to set a precedent of using native machine readable capabilities rather than playing regex catchup:
> 
> ```shell
> ip -json -details -stats address
> ip -json -details -stats link show
> ip -json route show table all
> ```

This is way better. I assumed there must have been a more easily parsed method. I'll use this I think.

dchadwic-xilinx commented:
It seems that perl's JSON-parser (`perl-JSON`) must be installed separately. For the sfreport script, we cannot expect customer systems to have this installed.
I'm still leaning towards using the `-json` output anyway as it's leagues more readable... and then playing "regex catchup".

ibeecraft-amd commented:
So it will need testing in more detail but since perl 5.14 we can use JSON:PP in scripts.

> JSON::PP has been in the Perl core since Perl 5.14, mainly for CPAN toolchain modules to parse META.json.

RHEL 7.9 has version 5.16 by default so I think we can safely use this package in sfreport with no issues:
```
[dellr630ag:perl]$cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.9 (Maipo)
[dellr630ag:perl]$perl --version

This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux-thread-multi
(with 41 registered patches, see perl -V for more detail)

Copyright 1987-2012, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.
```

Depending on speed and ease of use I may re-write my devlink patch to use this module for parsing.

KalepuAMD commented:
It's better for us to have a README or Release Notes file with an sfreport package that says this is the minimum Perl version required to run this version of sfreport, as we do have very few customers who use earlier versions of RHEL-7.9.

dchadwic-xilinx commented:
This now has no dependency on `JSON::PP`. Customers with an older version will simply get the 'print_preformatted` version.

pemberso-xilinx commented:
Can you split this up into multiple PRs?  It is adding backup tools, doing things with JSON and changing the appearance of the routing table.  I'd much prefer to review these separately.

(I'm not a fan of the new routing table with all N/A as I think much harder to read and I'd like to consider this separately)

dchadwic-xilinx commented:
@pemberso-xilinx 
I've removed the json-y bits for now and will raise a new PR once this one is approved and merged. I've reverted back to `print_preformatted for now

andresa-xilinx commented on sfreport.pl:
```diff
@@ -449,6 +451,7 @@ sub tabulate {
   }
 }
 
+
```
delete extra new line?

andresa-xilinx commented on sfreport.pl:
```diff
@@ -1946,18 +1958,27 @@ sub print_net_status {
-                    print_preformatted($report);
-                    last;
-                }
+        if (my $test = new FileHandle('ifconfig 2>&1 |')) {
```
`$test` filehandle is left opened, I believe it should be closed - possibly after `print_footer`?

dchadwic-xilinx replied:
Closing of the file handles is not done elsewhere in the file - and I assume this is auto cleaned up when the perl script finishes.
No harm in starting the correct style though.
